File: HomeAutomation/home-automation/src/main/java/observers/MotionDetector.java
package observers;

import models.EnvironmentState;

public class MotionDetector implements Observer {
    private boolean motionDetected;

    @Override
    public void update(EnvironmentState environmentState){
        this.motionDetected = environmentState.isMotionDetected();
        if (motionDetected) {
            System.out.println("Motion detected!");
        } else {
            System.out.println("No motion detected.");
        }
    }
}

File: HomeAutomation/HomeAutomation/Componenets/MotionDetector.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using HomeAutomation.Core;

namespace HomeAutomation.Componenets
{
    public class MotionDetector : IEnvironmentObserver
    {
        private List<string> motionLog = new();

        public void OnEnvironmentStateChanged(EnvironmentState state)
        {
            if (state.MotionDetected)
            {
                string logEntry = $"Motion detected at {DateTime.Now}";
                motionLog.Add(logEntry);
                Console.WriteLine(logEntry);
            }
        }

        public List<string> GetMotionLogs()
        {
            return motionLog;
        }
    }
}
